### PR TITLE
fixes #77

### LINKED ISSUE #77: Composition of unitary channels
Following code 
```
using QuantumInformation
u = UnitaryChannel([cos(1) sin(1); -sin(1) cos(1)])
compose(u, u)
```
results in
```
ERROR: MethodError: no method matching compose(::UnitaryChannel{Array{Float64,2}}, ::UnitaryChannel{Array{Float64,2}})
Closest candidates are:
  compose(::Type{T<:AbstractQuantumOperation{M<:(AbstractArray{#s52,2} where #s52<:Number)}}, ::T1<:AbstractQuantumOperation{M1<:(AbstractArray{#s38,2} where #s38<:Number)}, ::T2<:AbstractQuantumOperation{M2<:(AbstractArray{#s37,2} where #s37<:Number)}) where {M<:(AbstractArray{#s52,2} where #s52<:Number), M1<:(AbstractArray{#s38,2} where #s38<:Number), M2<:(AbstractArray{#s37,2} where #s37<:Number), T<:AbstractQuantumOperation{M<:(AbstractArray{#s52,2} where #s52<:Number)}, T1<:AbstractQuantumOperation{M1<:(AbstractArray{#s38,2} where #s38<:Number)}, T2<:AbstractQuantumOperation{M2<:(AbstractArray{#s37,2} where #s37<:Number)}} at /home/adam/.julia/packages/QuantumInformation/06n6F/src/channels/compositions.jl:79
Stacktrace:
 [1] top-level scope at none:0
```
And following
```
compose(UnitaryChannel{Array{Float64,2}}, u, u)
```
with 
```
ERROR: MethodError: Cannot `convert` an object of type SuperOperator{Array{Float64,2}} to an object of type UnitaryChannel{Array{Float64,2}}
Closest candidates are:
  convert(::Type{T}, ::T) where T at essentials.jl:154
Stacktrace:
 [1] compose(::Type{UnitaryChannel{Array{Float64,2}}}, ::UnitaryChannel{Array{Float64,2}}, ::UnitaryChannel{Array{Float64,2}}) at /home/adam/.julia/packages/QuantumInformation/06n6F/src/channels/compositions.jl:86
 [2] top-level scope at none:0
```

It seems that composition is badly defined. For first issue the type assertion in `function compose(Φ1::T, Φ2::T) where {T<:AbstractQuantumOperation{<:Number}}` is bad (`AbstractMatrix` is missing), similarly for `Base.:*(Φ1::T, Φ2::T)`. For the second a separate function for unitary composition may be convenient.

### PR DESCRIPTION
**Pull request by pgawron**

### CODE REVIEW

**Comment by adamglos92** [comment]
Proposed changes fixes the `UnitaryChannel` case, but still does not fix general case because of the code below.
```
function compose(Φ1::T, Φ2::T) where {T<:AbstractQuantumOperation{<:Number}}
    compose(T, Φ1, Φ2)
end
```
Furthermore I still suggest to implement separate functions for various combination of composition of `IdentityChannel`and `UnitaryChannel`
I think I can propose some additional tests
```
id = IdentityChannel(2)
u1 = UnitaryChannel([cos(1) sin(1); -sin(1) cos(1)])
@test compose(u1, id)
@test compose(id, u1)
@test compose(id, id)

Φ = KrausOperators([[1 0; 0 0], [0 0; 0 1]])
Φsparse = KrausOperators([sparse([1 0; 0 0]), [0 0; 0 1]])
@test compose(Φ, Φ)
@test compose(Φ, Φsparse)
```
Not sure what behavior should be expected if Matrix types differs, taking into account the fixes for Unitary case both should pass.

File: src/channels/compositions.jl [hunk 1]
```diff
@@ -86,6 +86,20 @@ function compose(::Type{T}, Φ1::T1, Φ2::T2) where {T<:AbstractQuantumOperation
     convert(T, so)
 end
 
+function compose(Φ1::UnitaryChannel{M1}, Φ2::UnitaryChannel{M2}) where {M1<:AbstractMatrix{<:Number}, M2<:AbstractMatrix{<:Number}}
+    M = promote_type(M1, M2)
+    compose(UnitaryChannel{M}, Φ1, Φ2)
+end
+
+function compose(::Type{UnitaryChannel{M}}, Φ1::UnitaryChannel{M1}, Φ2::UnitaryChannel{M2}) where {M<:AbstractMatrix{<:Number}, M1<:AbstractMatrix{<:Number}, M2<:AbstractMatrix{<:Number}}
+    Φ1.odim == Φ2.idim ? () : throw(ArgumentError("Unitaries are incompatible"))
```
Comment (lpawela) [inline src/channels/compositions.jl]: Do we need this check? The matrix multiplication will fail either way

File: src/channels/compositions.jl [hunk 2]
```diff
@@ -86,6 +86,20 @@ function compose(::Type{T}, Φ1::T1, Φ2::T2) where {T<:AbstractQuantumOperation
...
+  
+    um1 = Φ1.matrix
+    um2 = Φ2.matrix
+    um = um2 * um1
```
Comment (lpawela) [inline src/channels/compositions.jl]: um = Φ2.matrix * Φ1.matrix

**Comment by lpawela** [comment]
@adamglos92 sparse matrices are currently NOT supported, use at your own risk.

File: src/channels/compositions.jl (hunk 1 above)
Comment (pgawron) [reply src/channels/compositions.jl]: Error should be explained precisely. Matrix multiplication error is too generic and difficult to debug.

File: src/channels/compositions.jl (hunk 2 above)
Comment (pgawron) [reply src/channels/compositions.jl]: Ok.

**Pull request merged**